### DISCUSSION TITLE
How can I remove a dependency again that was added with `on`?

### DISCUSSION THREAD

**Discussion by danieltroger** [solidjs/solid#1120] [Q&A] [answered]
So I use `on` to have a function run when a signal updates (with `{ defer: true }` to not get the initial update) and I want to remove/"off" it again, how can I do that? Do I have to hack together my own on with defer by using `createComputed` i.e. or is there a better way?

**Comment by atk** [comment]
I would probably use the argument/return loop in createEffect:
```js
createEffect((run) => {
  const value = dependency();
  if (run === 1) myFunc(value);
  return run + 1;
}, 0);
```

**Comment by ryansolid** [comment, marked as answer]
Yeah `on` is meant for fixed dependencies. They get passed in every execution as well as previous and previously mapped versions etc.. I am having trouble thinking of an `off` API where this doesn't get really messy fast.

You can use conditional dependencies as normal and maybe in combination with `untrack`.  You can simulate deferred the way that @atk  showed above.

**Reply by danieltroger** [reply]
I wanted something that doesn't trigger on the initial value, like on with `{ defer: true }` and where I then can have it stop triggering when I want - with a nice API. All these counters are (IMO) a bit annoying because the if-statements are finicky to get to do exactly what I want and hard to debug.

But in both cases I actually ended up solving it differently anyways so maybe there's no need.

> I am having trouble thinking of an off API where this doesn't get really messy fast.

What about just providing an `off` function to the callback?
So changing `fn: (input: T, prevInput: T, prevValue?: U) => U,` to `fn: (input: T, prevInput: T, prevValue?: U, off?: () => void) => U,`

**Reply by ryansolid** [reply]
I guess there is no indicator what is being tracked now at any given time. Like with normal reactivity with the given inputs you know what was tracked on the last run based on its input. With `on` it is explicit. With `off` it's imperative, you can't really make any determination without knowing the full history of all inputs. If you noticed there is no `unsubscribe` in Solid or `once` and it's the same sort of problem. This sense of time isn't well represented, we want rules. In a sense reactivity is like circuits, which are only ever interested in the steady state. Programming for circuits and digital logic has this sort of declarative property that all decision points are able to introspected from the description of the behavior.

Thinking about this more I think we support this. You could put the logic in `on` argument.
```js
createEffect(on(() => {
  // This runs every time(including the first time) so you could mess around in here.
  // Just keep in mind that it will always run the last time that you remove it, returning what you what you want to pass into the function
}, fn))
```
In a sense this isn't much different than not using `on` with untrack but I suppose it handles the run once logic for you.
